### DISCUSSION TITLE
Program Chat Conversation Persistence?

### DISCUSSION THREAD

**Discussion by rickbraddy-pharma** [guidance-ai/guidance#192] [Q&A] [answered]
When running a program multiple times, is there a way to retain or pass conversational memory across program instantiations or executions?  I need to call the same program multiple times, having it continue an existing conversation, not seeing an example of this in the project notebooks (or perhaps I missed it).

**Comment by rickbraddy-pharma** [comment]
Figured out a way to do it...  here's a working example below. The key is to track the conversations and keep passing the conversational history back into the program while it's waiting.

import guidance

llm = guidance.llms.OpenAI("gpt-3.5-turbo")
guidance.llms.OpenAI.cache.clear()

prog = guidance(
'''{{#system~}}
You are a helpful assistant
{{~/system}}
{{~#geneach 'conversation' stop=False}}
{{#user~}}
{{convos}}
{{set 'this.user_text' (await 'user_text')}}
{{~/user}}
{{#assistant~}}
{{gen 'this.ai_text' temperature=0 max_tokens=300}}
{{~/assistant}}
{{~/geneach}}''', llm)

prompt = prog(convos=[], user_text = 'Who was first US President?')
convos = prompt['conversation']
prompt = prog(convos=convos, user_text = "When was he elected?")
convos.append(prompt['conversation'])
prompt = prog(convos=convos, user_text = "When did he die?")
convos.append(prompt['conversation'])
print(convos)

Output:
system
You are a helpful assistant
user
[{'user_text': 'Who was first US President?', 'ai_text': 'The first US President was George Washington.'}, {}, [{'user_text': 'When was he elected?', 'ai_text': 'George Washington was first elected as the President of the United States on April 30, 1789.'}, {}]]
When did he die?
assistant
George Washington died on December 14, 1799.
user
[{'user_text': 'Who was first US President?', 'ai_text': 'The first US President was George Washington.'}, {}, [{'user_text': 'When was he elected?', 'ai_text': 'George Washington was first elected as the President of the United States on April 30, 1789.'}, {}]]
{{set 'conversation[-1].user_text' (await 'user_text')}}
{{#assistant~}}
{{gen 'conversation[-1].ai_text' temperature=0 max_tokens=300}}
{{~/assistant}}
{{~#geneach 'conversation' stop=False}}
{{#user~}}
{{convos}}
{{set 'this.user_text' (await 'user_text')}}
{{~/user}}
{{#assistant~}}
{{gen 'this.ai_text' temperature=0 max_tokens=300}}
{{~/assistant}}
{{~/geneach}}
[{'user_text': 'Who was first US President?', 'ai_text': 'The first US President was George Washington.'}, {}, [{'user_text': 'When was he elected?', 'ai_text': 'George Washington was first elected as the President of the United States on April 30, 1789.'}, {}], [{'user_text': 'When did he die?', 'ai_text': 'George Washington died on December 14, 1799.'}, {}]]

**Reply by rickbraddy-pharma** [reply]
Not sure if this is the best or "recommended" way to pass conversational history around or not, but it's working for now.

**Comment by chrisfentiman** [comment, marked as answer]
@rickbraddy-pharma - it's up to you what is best based on your application needs.

For example, we have a chat application that creates context per request, so instead of having a long-running conversation in one thread, we do a linear execution of the prompt and dynamically fill in the chat context. Each new message is stored, and we build the chat history in context when the next request comes in. This enables you to build robust chat applications. However, this is just a chat example. The great thing about guidance is that you're completely in control and need to find the best way to accomplish your goal.
